File: cpp/tests/test_consume_thread_pool.cpp
// 消费线程弹性单测（对应 Java DefaultMQPushConsumer 的线程池配置与 updateCorePoolSize）。
//
// 为什么单独一个文件：Java 的 ThreadPoolExecutor 在 LinkedBlockingQueue（无界）下，
// **真实并发度 == corePoolSize**（max 永远用不到），所以 updateCorePoolSize 是"运行时能改
// 并发度"的 API。C++ 侧此前 POP 路径是"每批一个 detached 线程"（无上限）且
// setConsumeThreadNums() 完全没作用 —— 换成有界 core/max 执行器后，本文件把语义钉死。
//
// 另一个必须钉住的反直觉事实：**Java 5.5.1 的自动弹性（inc/decCorePoolSize）是空实现**，
// adjustThreadPool() 整套是 no-op。我们照抄 no-op，不允许"顺手修好"。
#include <atomic>
#include <chrono>
#include <cstdio>
#include <memory>
#include <string>
#include <thread>
#include <vector>

#include "rocketmq/client/consume_executor.h"
#include "rocketmq/client/consumer.h"
#include "rocketmq/common/message.h"
#include "rocketmq/common/message_const.h"

using namespace rocketmq;

namespace {

int fails = 0;
int checks = 0;

void expect(bool ok, const std::string& name, const std::string& detail = "") {
    ++checks;
    if (!ok) {
        ++fails;
        std::printf("FAIL %s %s\n", name.c_str(), detail.c_str());
    }
}

void expectEq(long long actual, long long expected, const std::string& name) {
    ++checks;
    if (actual != expected) {
        ++fails;
        std::printf("FAIL %s (actual=%lld expected=%lld)\n", name.c_str(), actual, expected);
    }
}

void sleepMs(int ms) { std::this_thread::sleep_for(std::chrono::milliseconds(ms)); }

MessageExt makeMsg(int64_t queueOffset, const std::string& maxOffset) {
    MessageExt m;
    m.topic = "T";
    m.body = "x";
    m.setQueueOffset(queueOffset);
    if (!maxOffset.empty()) {
        m.putProperty(MessageConst::PROPERTY_MAX_OFFSET, maxOffset);
    }
    return m;
}

// ---------------------------------------------------------------- ConsumeExecutor

void testSpawnsUpToCoreThenQueues() {
    ConsumeExecutor ex(2, 8, 30.0);
    std::atomic<int> running{0};
    std::atomic<bool> release{false};
    auto block = [&running, &release]() {
        running.fetch_add(1);
        while (!release.load()) sleepMs(5);
        running.fetch_sub(1);
    };
    ex.submit(block);
    ex.submit(block);
    // 两个 worker 被占住；第 3、4 个任务只能排队（未到 core 时不会建线程）
    ex.submit([]() {});
    ex.submit([]() {});
    sleepMs(100);
    expectEq(ex.workerCount(), 2, "spawns up to core");
    expectEq(ex.queuedCount(), 2, "beyond core is queued");
    release.store(true);
    ex.shutdown(true);
    expectEq(running.load(), 0, "all tasks finished after shutdown(true)");
}

void testRaisingCoreSpawnsForQueuedTasks() {
    // Java setCorePoolSize 的启发式：k = min(delta, 队列长度)，逐个补线程，队列空则停。
    ConsumeExecutor ex(1, 8, 30.0);
    std::atomic<bool> release{false};
    ex.submit([&release]() { while (!release.load()) sleepMs(5); });
    sleepMs(50);
    for (int i = 0; i < 3; ++i) ex.submit([]() {});
    expectEq(ex.workerCount(), 1, "only core thread before raising");
    expectEq(ex.queuedCount(), 3, "3 tasks queued");
    ex.setCorePoolSize(4);   // delta=3, queue=3 → 应补 3 个
    sleepMs(150);
    expectEq(ex.workerCount(), 4, "delta workers added on core raise");
    expectEq(ex.queuedCount(), 0, "queue drained");
    release.store(true);
    ex.shutdown(true);
}

void testExtraThreadRetiresCoreThreadDoesNot() {
    // > core 的线程空闲到 keepAlive 退出；<= core 的线程永不退出（Java allowCoreThreadTimeOut=false）
    ConsumeExecutor ex(1, 4, 0.2);
    ex.submit([]() {});
    ex.setCorePoolSize(2);   // 队列空 → 不补线程，core 变 2
    sleepMs(50);
    expectEq(ex.getCorePoolSize(), 2, "core pool size applied");
    ex.submit([]() {});      // 把 worker 抬到 2（仍 <= core）
    sleepMs(50);
    expectEq(ex.workerCount(), 2, "two workers alive");
    ex.setCorePoolSize(1);   // core 降到 1 → 多出来的那个成为"超编"
    sleepMs(600);            // 超过 keepAlive
    expectEq(ex.workerCount(), 1, "extra worker retired after keep alive");
    sleepMs(400);
    expectEq(ex.workerCount(), 1, "core worker never retires");
    ex.shutdown(true);
}

void testTaskExceptionDoesNotKillWorker() {
    ConsumeExecutor ex(1, 2, 30.0);
    ex.submit([]() { throw std::runtime_error("boom"); });
    sleepMs(200);
    expectEq(ex.handlerExceptionCount(), 1, "exception counted");
    expectEq(ex.workerCount(), 1, "worker survived exception");
    std::atomic<bool> ran{false};
    ex.submit([&ran]() { ran.store(true); });
    sleepMs(200);
    expect(ran.load(), "same worker keeps serving");
    ex.shutdown(true);
}

void testSubmitAfterShutdownThrows() {
    ConsumeExecutor ex(1, 2);
    ex.shutdown(false);
    bool threw = false;
    try {
        ex.submit([]() {});
    } catch (const std::runtime_error&) {
        threw = true;
    }
    expect(threw, "submit after shutdown throws");
    ex.shutdown(true);
}

void testShutdownWaitDrainsQueue() {
    ConsumeExecutor ex(1, 2, 30.0);
    std::atomic<int> done{0};
    for (int i = 0; i < 5; ++i) {
        ex.submit([&done]() {
            sleepMs(20);
            done.fetch_add(1);
        });
    }
    ex.shutdown(true);
    expectEq(done.load(), 5, "shutdown(wait) drains queued tasks");
}

void testZeroCoreStillRunsTasks() {
    // Java execute 的兜底分支：入队后若 workerCount == 0 仍要补一个线程
    ConsumeExecutor ex(0, 2, 0.1);
    std::atomic<bool> ran{false};
    ex.submit([&ran]() { ran.store(true); });
    sleepMs(200);
    expect(ran.load(), "core=0 still executes tasks");
    ex.shutdown(true);
}

void testCorePoolSizeMutation() {
    ConsumeExecutor ex(3, 9, 30.0);
    expectEq(ex.getCorePoolSize(), 3, "initial core");
    expectEq(ex.getMaximumPoolSize(), 9, "initial max");
    ex.setCorePoolSize(5);
    expectEq(ex.getCorePoolSize(), 5, "core raised");
    ex.setCorePoolSize(2);
    expectEq(ex.getCorePoolSize(), 2, "core lowered");
    ex.setCorePoolSize(12);  // Java 允许 core > max（等价于把 max 抬到 core）
    expectEq(ex.getMaximumPoolSize(), 12, "max raised to core");
    ex.shutdown(true);
}

// ---------------------------------------------------------------- 消费者侧（不联网）

void testJavaDefaults() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    expectEq(c.getConsumeThreadMin(), 20, "consumeThreadMin default = 20");
    expectEq(c.getConsumeThreadMax(), 20, "consumeThreadMax default = 20");
    expectEq(c.getAdjustThreadPoolNumsThreshold(), 100000, "adjustThreadPoolNumsThreshold default");
    expectEq(c.getCorePoolSize(), 20, "corePoolSize default = consumeThreadMin");
}

void testUpdateCorePoolSizeGuards() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    // Java 用无界队列 ⇒ 真实并发度 == core；默认 max=20，故 core 只能往**下**调
    expect(c.updateCorePoolSize(15), "15 accepted (below default max)");
    expectEq(c.getCorePoolSize(), 15, "core updated to 15");
    expect(!c.updateCorePoolSize(0), "0 rejected");
    expect(!c.updateCorePoolSize(-1), "-1 rejected");
    expect(!c.updateCorePoolSize(20), "== consumeThreadMax rejected");
    expect(c.updateCorePoolSize(19), "19 accepted (just below max)");
    expectEq(c.getCorePoolSize(), 19, "core updated to 19");
    // Short.MAX_VALUE 上界：要把 consumeThreadMax 抬上去才轮得到这条守卫
    c.setConsumeThreadMax(40000);
    expect(!c.updateCorePoolSize(32768), "32768 rejected (above Short.MAX_VALUE)");
    expect(c.updateCorePoolSize(32767), "32767 accepted (upper bound itself)");
    expectEq(c.getCorePoolSize(), 32767, "core updated to Short.MAX_VALUE");
}

// 回归：默认 max 曾照抄 4.x 的 64，于是 20~63 这些 Java 会**忽略**的值能生效。
// Java 5.5.1 的 consumeThreadMax 与 min 同为 20（DefaultMQPushConsumer:169），守卫
// `corePoolSize < consumeThreadMax` 把默认配置下的上调全挡掉；差异不报错，只表现为
// "同一个 updateCorePoolSize(30)，本端口真的改了并发度、Java 没改"。
void testDefaultMaxIsJava5xTwenty() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    expectEq(c.getConsumeThreadMax(), 20, "default max is Java 5.x's 20");
    expect(!c.updateCorePoolSize(30), "30 rejected under default max");
    expect(!c.updateCorePoolSize(21), "21 rejected under default max");
    expectEq(c.getCorePoolSize(), 20, "nothing landed");
    // 抬 max 之后区间重新打开（setter 与 Java 一样是裸赋值，只夹 >= 1）
    c.setConsumeThreadMax(30);
    expect(c.updateCorePoolSize(25), "25 accepted after raising max to 30");
    expectEq(c.getCorePoolSize(), 25, "core updated to 25");
}

void testSetConsumeThreadNumsSetsBothAndCore() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    c.setConsumeThreadNums(4);
    expectEq(c.getConsumeThreadMin(), 4, "min set by setConsumeThreadNums");
    expectEq(c.getConsumeThreadMax(), 4, "max set by setConsumeThreadNums");
    expectEq(c.getCorePoolSize(), 4, "core follows setConsumeThreadNums");
    expect(!c.updateCorePoolSize(4), "4 rejected after max became 4");
    expect(c.updateCorePoolSize(3), "3 accepted after max became 4");
}

void testConsumeThreadMinSetterMovesCoreMaxDoesNot() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    c.setConsumeThreadMin(8);
    expectEq(c.getCorePoolSize(), 8, "core follows consumeThreadMin");
    c.setConsumeThreadMax(16);
    expectEq(c.getConsumeThreadMax(), 16, "max updated");
    expectEq(c.getCorePoolSize(), 8, "max change does not move core");
    c.setConsumeThreadMin(0);
    c.setConsumeThreadMax(-5);
    expectEq(c.getConsumeThreadMin(), 1, "min clamped to 1");
    expectEq(c.getConsumeThreadMax(), 1, "max clamped to 1");
}

void testMsgAccCntRules() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    // accTotal = MAX_OFFSET - queueOffset，取**最后一条**
    c.updateMsgAccCnt("k1", {makeMsg(10, "100"), makeMsg(12, "100")});
    expectEq(c.msgAccCnt("k1"), 88, "acc = maxOffset - lastQueueOffset");
    // 非正不更新
    c.updateMsgAccCnt("k2", {makeMsg(100, "100")});
    expectEq(c.msgAccCnt("k2"), 0, "zero delta ignored");
    c.updateMsgAccCnt("k2", {makeMsg(200, "100")});
    expectEq(c.msgAccCnt("k2"), 0, "negative delta ignored");
    // 属性缺失 / 非数字 / 带尾巴一律忽略
    c.updateMsgAccCnt("k3", {makeMsg(10, "")});
    expectEq(c.msgAccCnt("k3"), 0, "missing property ignored");
    c.updateMsgAccCnt("k3", {makeMsg(10, "not-a-number")});
    expectEq(c.msgAccCnt("k3"), 0, "garbage property ignored");
    c.updateMsgAccCnt("k3", {makeMsg(10, "100abc")});
    expectEq(c.msgAccCnt("k3"), 0, "trailing garbage ignored");
    // 新一批覆盖旧值
    c.updateMsgAccCnt("k4", {makeMsg(10, "500")});
    c.updateMsgAccCnt("k4", {makeMsg(400, "450")});
    expectEq(c.msgAccCnt("k4"), 50, "latest batch overwrites");
    // 求和：换一个干净的消费者，避免被上面几组 key 的残留值影响
    DefaultMQPushConsumer sum("GID_ThreadPoolUnit2");
    sum.updateMsgAccCnt("a", {makeMsg(0, "30")});
    sum.updateMsgAccCnt("b", {makeMsg(0, "70")});
    expectEq(sum.msgAccCnt(), 100, "sum across queues");
    expectEq(sum.computeAccumulationTotal(), 100, "computeAccumulationTotal sums msgAccCnt");
}

void testAdjustThreadPoolIsNoOp() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    c.setAdjustThreadPoolNumsThreshold(100);
    c.updateMsgAccCnt("a", {makeMsg(0, "500")});   // 500 >= 100 → inc 分支
    const int32_t before = c.getCorePoolSize();
    c.adjustThreadPool();
    expectEq(c.getCorePoolSize(), before, "inc branch is a no-op");
    c.setAdjustThreadPoolNumsThreshold(1000);
    c.updateMsgAccCnt("a", {makeMsg(0, "10")});    // 10 < 800 → dec 分支
    c.adjustThreadPool();
    expectEq(c.getCorePoolSize(), before, "dec branch is a no-op");
    // 自动 no-op ≠ API 失效
    expect(c.updateCorePoolSize(11), "explicit update still works after adjustThreadPool");
    expectEq(c.getCorePoolSize(), 11, "explicit core applied");
}

void testExecutorObservabilityWithoutStart() {
    DefaultMQPushConsumer c("GID_ThreadPoolUnit");
    // 未 start（未建执行器）时观测值恒为 0，不应崩
    expectEq(c.consumeExecutorWorkers(), 0, "no executor before start");
    expectEq(c.consumeExecutorQueued(), 0, "no queued tasks before start");
}

}  // namespace

int main() {
    testSpawnsUpToCoreThenQueues();
    testRaisingCoreSpawnsForQueuedTasks();
    testExtraThreadRetiresCoreThreadDoesNot();
    testTaskExceptionDoesNotKillWorker();
    testSubmitAfterShutdownThrows();
    testShutdownWaitDrainsQueue();
    testZeroCoreStillRunsTasks();
    testCorePoolSizeMutation();
    testJavaDefaults();
    testUpdateCorePoolSizeGuards();
    testDefaultMaxIsJava5xTwenty();
    testSetConsumeThreadNumsSetsBothAndCore();
    testConsumeThreadMinSetterMovesCoreMaxDoesNot();
    testMsgAccCntRules();
    testAdjustThreadPoolIsNoOp();
    testExecutorObservabilityWithoutStart();

    std::printf("%s: %d checks, %d failed\n", fails == 0 ? "PASS" : "FAIL", checks, fails);
    return fails == 0 ? 0 : 1;
}

File: python/tests/test_consume_thread_pool.py
# -*- coding: utf-8 -*-
"""消费线程弹性单测（对应 Java DefaultMQPushConsumer 的线程池配置与 updateCorePoolSize）。

为什么需要它：Java 的 ``ThreadPoolExecutor`` 在 ``LinkedBlockingQueue``（无界）下，
**真实并发度 == corePoolSize**（``max`` 永远用不到），所以 ``updateCorePoolSize`` 是
"运行时真的能改并发度"的 API，而标准的 ``concurrent.futures.ThreadPoolExecutor``
**没有 core 这个概念** —— 用它就等于把这个能力做没了。因此本项目自己实现
``ConsumeExecutor``（core/max 两档），本文件把它的语义和 Java 守卫逐条钉住。

另外钉住一个反直觉的事实：**Java 5.5.1 的自动弹性（inc/decCorePoolSize）是空实现**，
``adjustThreadPool()`` 整套调度是 no-op。我们照抄 no-op，不允许"顺手修好"。

覆盖：
  - ConsumeExecutor：core 内建线程 / 超 core 入队 / 提升 core 补线程 /
    超编线程空闲退出而 core 线程不退 / 任务异常不杀 worker / shutdown 语义
  - 消费者：Java 默认值、update_core_pool_size 的五条守卫边界、get_core_pool_size
  - msgAccCnt（MAX_OFFSET - queueOffset）与 compute_accumulation_total
  - adjust_thread_pool 的 inc/dec 确实是 no-op（改阈值也不动 core）
  - 307 运行信息里的 PROP_THREADPOOL_CORE_SIZE 取的是 core pool size
"""
from __future__ import annotations

import threading
import time

import pytest

from rocketmq.client.consume_executor import ConsumeExecutor
from rocketmq.client.consumer import DefaultMQPushConsumer
from rocketmq.common.message import MessageExt, MessageQueue
from rocketmq.common.message_const import MessageConst
from rocketmq.remoting.protocol.body import ConsumerRunningInfo

GROUP = "GID_ThreadPoolUnit"


# ------------------------------------------------------------------ ConsumeExecutor

class TestConsumeExecutor:
    def test_spawns_up_to_core_then_queues(self):
        """core=2：前两个任务各起一个线程；第 3 个任务只入队，不再建线程。"""
        ex = ConsumeExecutor(2, 8, keep_alive_seconds=30)
        started = threading.Event()
        release = threading.Event()

        def block(tag):
            started.set()
            release.wait(5)

        try:
            ex.submit(block, "a")
            assert started.wait(3)
            started.clear()
            ex.submit(block, "b")
            assert started.wait(3)
            # 两个 worker 都被占住，第 3 个任务只能在队列里等
            ex.submit(block, "c")
            time.sleep(0.1)
            assert ex.worker_count() == 2
            assert ex.queued_count() == 1
        finally:
            release.set()
            ex.shutdown(wait=True)

    def test_raising_core_spawns_for_queued_tasks(self):
        """set_core_pool_size 变大且队列非空 → 立刻补足 worker（Java setCorePoolSize）。"""
        ex = ConsumeExecutor(1, 8, keep_alive_seconds=30)
        release = threading.Event()
        first_started = threading.Event()

        def block():
            first_started.set()
            release.wait(5)

        try:
            ex.submit(block)
            assert first_started.wait(3)
            for _ in range(3):
                ex.submit(lambda: None)
            assert ex.worker_count() == 1
            assert ex.queued_count() == 3
            ex.set_core_pool_size(4)
            deadline = time.time() + 3
            while ex.worker_count() < 4 and time.time() < deadline:
                time.sleep(0.02)
            assert ex.worker_count() == 4
        finally:
            release.set()
            ex.shutdown(wait=True)

    def test_extra_thread_retires_after_keep_alive_core_thread_does_not(self):
        """> core 的线程空闲到 keep_alive 就退出；<= core 的线程永不退出。"""
        ex = ConsumeExecutor(1, 4, keep_alive_seconds=0.2)
        try:
            ex.submit(lambda: None)
            ex.set_core_pool_size(2)          # 队列空 → 不补线程，core 现在是 2
            time.sleep(0.05)
            assert ex.get_core_pool_size() == 2
            # 此刻只有 1 个 worker；再提交任务把它抬到 2（仍 <= core，应在超时后存活）
            ex.submit(lambda: None)
            time.sleep(0.05)
            assert ex.worker_count() == 2
            ex.set_core_pool_size(1)          # core 降到 1 → 多出的那个属于"超编"
            time.sleep(0.6)                   # 超过 keep_alive
            assert ex.worker_count() == 1     # 超编线程已退出
            time.sleep(0.4)
            assert ex.worker_count() == 1     # core 内线程不会退出
        finally:
            ex.shutdown(wait=True)

    def test_task_exception_does_not_kill_worker(self):
        ex = ConsumeExecutor(1, 2, keep_alive_seconds=30)
        done = threading.Event()
        try:
            def boom():
                raise ValueError("boom")

            ex.submit(boom)
            time.sleep(0.2)
            assert ex.handler_exception_count() == 1
            assert ex.worker_count() == 1
            ex.submit(done.set)
            assert done.wait(3)               # 同一个 worker 还能继续干活
        finally:
            ex.shutdown(wait=True)

    def test_submit_after_shutdown_raises(self):
        ex = ConsumeExecutor(1, 2)
        ex.shutdown()
        with pytest.raises(RuntimeError):
            ex.submit(lambda: None)

    def test_shutdown_wait_drains_queued_tasks(self):
        """Java shutdown() 不丢已提交任务：wait=True 要等队列跑完。"""
        ex = ConsumeExecutor(1, 2, keep_alive_seconds=30)
        counter = {"n": 0}
        lock = threading.Lock()

        def work():
            time.sleep(0.02)
            with lock:
                counter["n"] += 1

        for _ in range(5):
            ex.submit(work)
        ex.shutdown(wait=True)
        assert counter["n"] == 5

    def test_zero_core_means_run_to_completion_style_queue(self):
        """core=0（不建常驻线程）时任务仍然会被执行（提交时按需起线程）。"""
        ex = ConsumeExecutor(0, 2, keep_alive_seconds=0.1)
        done = threading.Event()
        try:
            ex.submit(done.set)
            assert done.wait(3)
        finally:
            ex.shutdown(wait=True)


# ------------------------------------------------------------------ 消费者侧

class TestConsumerDefaultsAndGuards:
    def test_java_defaults(self):
        c = DefaultMQPushConsumer(GROUP)
        assert c.get_consume_thread_min() == 20          # Java consumeThreadMin 默认 20 (:162)
        assert c.get_consume_thread_max() == 20          # Java consumeThreadMax 默认 20 (:169)
        assert c.get_adjust_thread_pool_nums_threshold() == 100000
        assert c.get_core_pool_size() == 20              # = consumeThreadMin

    def test_update_core_pool_size_happy_path(self):
        """Java 用无界队列 ⇒ 真实并发度 == core；默认配置下 core 只能往**下**调。"""
        c = DefaultMQPushConsumer(GROUP)
        assert c.update_core_pool_size(15) is True
        assert c.get_core_pool_size() == 15

    def test_update_core_pool_size_rejects_zero_and_negative(self):
        c = DefaultMQPushConsumer(GROUP)
        assert c.update_core_pool_size(0) is False
        assert c.update_core_pool_size(-1) is False
        assert c.get_core_pool_size() == 20              # 未生效

    def test_update_core_pool_size_rejects_above_short_max(self):
        """Java 守卫 corePoolSize <= Short.MAX_VALUE(32767)。

        注意两道守卫是 **与** 关系：要让 Short.MAX_VALUE 成为实际卡点，必须先把
        consumeThreadMax 抬到 32767 以上，否则先被 ``n < consumeThreadMax`` 拒掉。
        """
        c = DefaultMQPushConsumer(GROUP)
        c.set_consume_thread_max(40000)
        assert c.update_core_pool_size(32768) is False
        assert c.update_core_pool_size(32767) is True    # 上界本身是合法的
        assert c.get_core_pool_size() == 32767

    def test_update_core_pool_size_rejects_equal_to_consume_thread_max(self):
        """Java 守卫 corePoolSize < consumeThreadMax —— **等于**也不行。"""
        c = DefaultMQPushConsumer(GROUP)
        assert c.update_core_pool_size(20) is False      # == consumeThreadMax（5.x 默认）
        assert c.update_core_pool_size(19) is True
        assert c.get_core_pool_size() == 19

    def test_default_max_is_java_5x_twenty_not_four_x_sixty_four(self):
        """回归：默认 max 曾照抄 4.x 的 64，于是 20~63 这些 Java 会**忽略**的值能生效。

        Java 5.5.1 的 consumeThreadMax 与 min 同为 20（DefaultMQPushConsumer:169），
        守卫 ``n < consumeThreadMax`` 因此把默认配置下的上调全挡掉 —— 差异不会报错，
        只表现为"同一个 update_core_pool_size(30)，本端口真的改了并发度、Java 没改"。
        """
        c = DefaultMQPushConsumer(GROUP)
        assert c.get_consume_thread_max() == 20
        assert c.update_core_pool_size(30) is False
        assert c.update_core_pool_size(21) is False
        assert c.get_core_pool_size() == 20              # 一个都没落下去
        # 抬 max 之后区间重新打开（setter 与 Java 一样是裸赋值，只夹 >= 1）
        c.set_consume_thread_max(30)
        assert c.update_core_pool_size(25) is True
        assert c.get_core_pool_size() == 25

    def test_set_consume_thread_nums_sets_min_max_and_core(self):
        c = DefaultMQPushConsumer(GROUP)
        c.set_consume_thread_nums(4)
        assert c.get_consume_thread_min() == 4
        assert c.get_consume_thread_max() == 4
        assert c.get_core_pool_size() == 4
        # max 现在是 4 → 守卫按新 max 判定
        assert c.update_core_pool_size(4) is False
        assert c.update_core_pool_size(3) is True

    def test_consume_thread_min_setter_moves_core(self):
        c = DefaultMQPushConsumer(GROUP)
        c.set_consume_thread_min(8)
        assert c.get_consume_thread_min() == 8
        assert c.get_core_pool_size() == 8
        c.set_consume_thread_max(16)
        assert c.get_consume_thread_max() == 16
        assert c.get_core_pool_size() == 8               # 改 max 不动 core

    def test_sizes_are_clamped_to_at_least_one(self):
        c = DefaultMQPushConsumer(GROUP)
        c.set_consume_thread_min(0)
        c.set_consume_thread_max(-5)
        assert c.get_consume_thread_min() == 1
        assert c.get_consume_thread_max() == 1

    def test_threshold_setter(self):
        c = DefaultMQPushConsumer(GROUP)
        c.set_adjust_thread_pool_nums_threshold(0)
        assert c.get_adjust_thread_pool_nums_threshold() == 0

    def test_executor_is_owned_so_guard_passes(self):
        c = DefaultMQPushConsumer(GROUP)
        assert c._owns_consume_executor is True
        assert c.update_core_pool_size(15) is True


# ------------------------------------------------------------------ msgAccCnt / 阈值

def _msg(queue_offset: int, max_offset=None) -> MessageExt:
    m = MessageExt(topic="T", body=b"x")
    m.queue_offset = queue_offset
    if max_offset is not None:
        m.properties[MessageConst.PROPERTY_MAX_OFFSET] = str(max_offset)
    return m


class TestMsgAccCnt:
    def test_acc_cnt_is_max_offset_minus_queue_offset(self):
        """Java ProcessQueue:148-158 —— accTotal = MAX_OFFSET - queueOffset，取最后一条。"""
        c = DefaultMQPushConsumer(GROUP)
        c._update_msg_acc_cnt("k1", [_msg(10, 100), _msg(12, 100)])
        assert c.msg_acc_cnt("k1") == 88                 # 100 - 12

    def test_acc_cnt_ignores_non_positive(self):
        c = DefaultMQPushConsumer(GROUP)
        c._update_msg_acc_cnt("k1", [_msg(100, 100)])     # == 0 → 不更新
        assert c.msg_acc_cnt("k1") == 0
        c._update_msg_acc_cnt("k1", [_msg(200, 100)])     # < 0 → 不更新
        assert c.msg_acc_cnt("k1") == 0

    def test_acc_cnt_missing_property_is_ignored(self):
        c = DefaultMQPushConsumer(GROUP)
        c._update_msg_acc_cnt("k1", [_msg(10)])           # 没有 MAX_OFFSET 属性
        assert c.msg_acc_cnt("k1") == 0

    def test_acc_cnt_garbage_property_is_ignored(self):
        c = DefaultMQPushConsumer(GROUP)
        m = _msg(10)
        m.properties[MessageConst.PROPERTY_MAX_OFFSET] = "not-a-number"
        c._update_msg_acc_cnt("k1", [m])
        assert c.msg_acc_cnt("k1") == 0

    def test_acc_cnt_keeps_largest_snapshot_of_last_pull(self):
        c = DefaultMQPushConsumer(GROUP)
        c._update_msg_acc_cnt("k1", [_msg(10, 500)])
        c._update_msg_acc_cnt("k1", [_msg(400, 450)])     # 新的一批，积压变小
        assert c.msg_acc_cnt("k1") == 50

    def test_compute_accumulation_total_sums_all_queues(self):
        c = DefaultMQPushConsumer(GROUP)
        c._update_msg_acc_cnt("a", [_msg(0, 30)])
        c._update_msg_acc_cnt("b", [_msg(0, 70)])
        assert c.compute_accumulation_total() == 100
        assert c.msg_acc_cnt() == 100


class TestAdjustThreadPoolIsNoOp:
    def test_inc_branch_does_not_change_core_pool_size(self):
        """acc >= 1.0 * threshold → Java 调 incCorePoolSize()，而它是空实现。"""
        c = DefaultMQPushConsumer(GROUP)
        c.set_adjust_thread_pool_nums_threshold(100)
        c._update_msg_acc_cnt("a", [_msg(0, 500)])        # 500 >= 100
        before = c.get_core_pool_size()
        c.adjust_thread_pool()
        assert c.get_core_pool_size() == before           # no-op

    def test_dec_branch_does_not_change_core_pool_size(self):
        """acc < 0.8 * threshold → Java 调 decCorePoolSize()，同样是空实现。"""
        c = DefaultMQPushConsumer(GROUP)
        c.set_adjust_thread_pool_nums_threshold(1000)
        c._update_msg_acc_cnt("a", [_msg(0, 10)])         # 10 < 800
        before = c.get_core_pool_size()
        c.adjust_thread_pool()
        assert c.get_core_pool_size() == before           # no-op

    def test_explicit_update_still_works_after_adjust(self):
        """自动弹性 no-op ≠ API 失效：显式 update_core_pool_size 仍然生效。"""
        c = DefaultMQPushConsumer(GROUP)
        c._update_msg_acc_cnt("a", [_msg(0, 10 ** 6)])
        c.adjust_thread_pool()
        assert c.update_core_pool_size(11) is True
        assert c.get_core_pool_size() == 11


class TestRunningInfoReflectsCorePoolSize:
    def test_prop_threadpool_core_size_uses_core_not_max(self):
        c = DefaultMQPushConsumer(GROUP)
        c.name_server_addrs = ["127.0.0.1:9876"]
        c.namespace = None
        c._start_time = time.time()
        c.update_core_pool_size(13)
        info = c.consumer_running_info()
        assert info.properties[ConsumerRunningInfo.PROP_THREADPOOL_CORE_SIZE] == "13"
        assert info.properties[ConsumerRunningInfo.PROP_CONSUME_TYPE] == "CONSUME_PASSIVELY"

    def test_prop_follows_set_consume_thread_nums(self):
        c = DefaultMQPushConsumer(GROUP)
        c.name_server_addrs = []
        c.namespace = None
        c._start_time = time.time()
        c.set_consume_thread_nums(7)
        info = c.consumer_running_info()
        assert info.properties[ConsumerRunningInfo.PROP_THREADPOOL_CORE_SIZE] == "7"
